src/providers/hana/qgshanacrsutils.cpp

```diff
@@ -0,0 +1,111 @@
+/***************************************************************************
+   qgshanacrsutils.cpp
+   --------------------------------------
+   Date      : 13-10-2019
+   Copyright : (C) SAP SE
+   Author    : Maxim Rylov
+ ***************************************************************************/
+
+/***************************************************************************
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ ***************************************************************************/
+#include "qgshanacrsutils.h"
+#include "qgshanaexception.h"
+
+#if PROJ_VERSION_MAJOR>=6
+#include <proj.h>
+#include "qgsprojutils.h"
+#else
+#include <ogr_srs_api.h>
+#endif
+
+double QgsHanaCrsUtils::getAngularUnits( const QgsCoordinateReferenceSystem &crs )
+{
+  auto throwUnableToGetAngularUnits = []()
+  {
+    throw QgsHanaException( "Unable to retrieve angular units from a spatial reference system" );
+  };
+
+#if PROJ_VERSION_MAJOR>=6
+  PJ *pjCrs = crs.projObject();
+  if ( !pjCrs )
+    throwUnableToGetAngularUnits();
+  PJ_CONTEXT *context = QgsProjContext::get();
+  QgsProjUtils::proj_pj_unique_ptr pjCoordinateSystem( proj_crs_get_coordinate_system( context, pjCrs ) );
+  if ( !pjCoordinateSystem )
+    throwUnableToGetAngularUnits();
+
+  const int axisCount = proj_cs_get_axis_count( context, pjCoordinateSystem.get() );
+  if ( axisCount <= 0 )
+    throwUnableToGetAngularUnits();
+
+  double factor;
+  bool ret = proj_cs_get_axis_info( context, pjCoordinateSystem.get(), 0,
+                                    nullptr,
+                                    nullptr,
+                                    nullptr,
+                                    &factor,
+                                    nullptr,
+                                    nullptr,
+                                    nullptr );
+  if ( !ret )
+    throwUnableToGetAngularUnits();
+  return factor;
+#else
+  OGRSpatialReferenceH hCRS = OSRNewSpatialReference( nullptr );
+  int errcode = OSRImportFromProj4( hCRS, crs.toProj().toUtf8() );
+  if ( errcode != OGRERR_NONE )
+  {
+    if ( hCRS )
+      OSRRelease( hCRS );
+    throwUnableToGetAngularUnits();
+  }
+
+  char *angularUnits = nullptr;
+  double factor = OSRGetAngularUnits( hCRS, &angularUnits );
+  OSRRelease( hCRS );
+  return factor;
+#endif
+}
+
+bool QgsHanaCrsUtils::identifyCrs( const QgsCoordinateReferenceSystem &crs, QString &name, long &srid )
```

Comment:
Why is all this needed? It looks like duplication of logic which is already present in QgsCoordinateReferenceSystem

Reply:
Do you only refer to the `identifyCrs` method or also to the `getAngularUnits` method?

I guess we could replace the implementation of `identifyCrs` by a call to `QgisCoordinateReferenceSystem::authid()`, but I did not find a method that could replace `getAngularUnits`. Or did I miss something?

Reply:
I can't see anywhere that getAngularUnits is actually called though -- am I missing something? But otherwise yes, identify CRS looks like a reimplementation of QgsCoordinateReferenceSystem::authid() and should be removed. (But be aware that the code part of an authid is not always numeric -- proj_get_id_code will sometimes return non-numeric strings.

Reply:
I've changed the code to use `QgsCoordinateReferenceSystem::authid()` in 55e146f.

We call method `getAngularUnits()` in `qgshanafeatureiterator.cpp` to clamp the filter rectangle. In geographic spatial reference systems, HANA accepts only coordinates from the "normalized" range. If the unit is degree, the longitude must be in `[-180,180]` and the latitude in `[-90,90]`.

However, in QGIS geographic spatial reference system are still projected and you can draw filter rectangles that have coordinates outside the "normalized" range. So, we have to clamp the filter rectangle.

Unfortunately, not all geographic spatial reference systems use degree as angular unit, e.g. EPSG:4807 uses grad instead. Hence, we need the conversion factor of the angular unit used in the spatial reference system to radian.

Should we move this functionality to `QgsCoordinateReferenceSystem` instead?